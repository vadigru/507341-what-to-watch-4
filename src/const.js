export const MOVIES_DEFAULT_AMOUNT = 8;
export const MOVIES_STEP_AMOUNT = 8;
export const MAX_GENRES_AMOUNT = 9;
export const ALL_GENRES = `All genres`;
export const START_DELAY = 1000;

export const Tab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const DefaultTab = {
  GENRE: `All genres`,
  MOVIE: `Overview`
};

export const RatingLevel = {
  ZERO: 0,
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  AWESOME: 10
};

export const AppRoute = {
  ROOT: `/`,
  MOVIE_PAGE: `/movie`,
  SIGN_IN: `/login`,
  ADD_REVIEW: `/review`,
};
