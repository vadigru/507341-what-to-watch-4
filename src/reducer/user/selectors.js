import Namespace from "../namespace.js";

export const getAuthorizationStatus = (state) => {
  return state[Namespace.USER].authorizationStatus;
};

export const getInvalidAuthorizationStatus = (state) => {
  return state[Namespace.USER].isValidAuthorization;
};

export const getAvatar = (state) => {
  return state[Namespace.USER].avatarUrl;
};
