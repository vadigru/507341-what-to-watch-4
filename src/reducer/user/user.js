import {extend} from "../../utils/common.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`
};

const SERVER_URL = `https://4.react.pages.academy`;

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  isValidAuthorization: true,
  avatarUrl: ``,
};

const ActionType = {
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  IS_INVALID_AUTHORIZATION: `IS_INVALID_AUTHORIZATION`,
  SET_AVATAR_URL: `SET_AVATAR_URL`
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status
  }),
  requireValidAuthorization: (isValid) => ({
    type: ActionType.IS_INVALID_AUTHORIZATION,
    payload: isValid
  }),
  setAvatarUrl: (avatarUrl) => ({
    type: ActionType.SET_AVATAR_URL,
    payload: avatarUrl
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRE_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload
      });
    case ActionType.IS_INVALID_AUTHORIZATION:
      return extend(state, {
        isValidAuthorization: action.payload
      });
    case ActionType.SET_AVATAR_URL:
      return extend(state, {
        avatarUrl: `${SERVER_URL}${action.payload}`
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err) => {
      throw err;
    });
  },

  login: (authData) => (dispatch, getState, api) => {
    // dispatch(ActionCreator.requireValidAuthorization(true));
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password
    })
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err)=>{
      dispatch(ActionCreator.requireValidAuthorization(false));
      throw err;
    });
  }
};

export {ActionCreator, ActionType, AuthorizationStatus, Operation, reducer};
