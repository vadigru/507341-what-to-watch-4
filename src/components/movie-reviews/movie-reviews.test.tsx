import * as React from "react";
import * as renderer from "react-test-renderer";
import MovieReviews from "./movie-reviews";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import Namespace from "../../reducer/namespace";
import {ALL_GENRES, MOVIES_DEFAULT_AMOUNT} from "../../const";
import {AuthorizationStatus} from "../../reducer/user/user";

const mockStore = configureStore([]);

const mock = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Kate Muir`
    },
    rating: 8.9,
    comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    date: `2019-05-08T14:13:56.569Z`
  }
];

it(`Should render MovieDetails component`, () => {
  const store = mockStore({
    [Namespace.DATA]: {
      films: [],
      promo: {},
      reviews: [],
      isFilmsLoading: false,
      isPromoLoading: false,
      isReviewsLoading: false,
      isReviewPosting: false,
      isReviewSendingError: false,
    },
    [Namespace.STATE]: {
      genre: ALL_GENRES,
      showedMovies: MOVIES_DEFAULT_AMOUNT,
      selectedMovieId: 0
    },
    [Namespace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      isValidAuthorization: true,
      avatarUrl: ``,
    },
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <MovieReviews
            loadingReviews={true}
            reviews={mock}
          />
        </Provider>
    )
  .toJSON();

  expect(tree).toMatchSnapshot();
});
