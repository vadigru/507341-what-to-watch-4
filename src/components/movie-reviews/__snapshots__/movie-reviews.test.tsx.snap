// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render MovieDetails component 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  />
  <div
    className="movie-card__reviews-col"
  />
</div>
`;
